Implement BrokerHostUsage using java

sschepens opened the pull request:
### Motivation

Issue [#81](https://github.com/yahoo/pulsar/issues/81)
I've made a first go a this, I believe it's working ok, please review :blush: 
### Modifications

I've removed support for usage script and made `BrokerHostUsage` calculate metrics from system data.
Current implementation only supports Linux, we should probably find a way to implement the same functionallity for other OS.
### Result

Broker can collect usage metrics without need for a script.

yahoocla commented:
CLA is valid!

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/BrokerHostUsage.java:
```diff
@@ -71,25 +91,152 @@ public BrokerHostUsage(PulsarService pulsar) {
-            if (exitCode != 0) {
-                LOG.warn("Process exited with non-zero exit code - [{}], stderr - [{}] ", exitCode, writer.toString());
-                throw new IOException(writer.toString());
+            String[] words = Files.lines(Paths.get("/proc/stat"))
```
This is a bit terse, we should include a comment with an example of the file line. 😄 
The computation looks correct though.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/BrokerHostUsage.java:
```diff
@@ -71,25 +91,152 @@ public BrokerHostUsage(PulsarService pulsar) {
+
+            return new CpuStat(values.sum(), values.limit(3).sum());
+        } catch (IOException e) {
+            return new CpuStat(0L, 0L);
```
We should not blindly ignore exception, otherwise it would be very difficult to discover any problem.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/BrokerHostUsage.java:
```diff
@@ -71,25 +91,152 @@ public BrokerHostUsage(PulsarService pulsar) {
+                    .map(path -> path.getFileName().toString())
+                    .collect(Collectors.toList());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Same here, we should either propagate the IOException or at least print a warning log message if we can't read the information.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/BrokerHostUsage.java:
```diff
@@ -71,25 +91,152 @@ public BrokerHostUsage(PulsarService pulsar) {
+            } catch (IOException e) {
+                return 0d;
+            }
+        }).sum() * 1024 / 8;
```
We should report `kbits/s` for NIC speed and traffic.

sschepens replied:
ok!

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/BrokerHostUsage.java:
```diff
@@ -71,25 +91,152 @@ public BrokerHostUsage(PulsarService pulsar) {
      */
-    public String getBrokerHostUsage() throws IOException {
-        StringWriter writer = new StringWriter();
+    public synchronized SystemResourceUsage getBrokerHostUsage() throws IOException {
```
The refresh should not be done every time this method is called, but rather at regular intervals (eg: every 1min)

sschepens replied:
If calculations are done every 1min, but the data es retreived on a greater period, does it really make sense to make calculations at shorter periods? Would we always return the average usage of the last minute?

sschepens replied:
Current implementation runs the script everytime this method is called.
How often is this method actually called?
This method would return the average usage between two calls.

merlimat commented on pulsar-broker/src/test/resources/test-script-pulsar-broker-host-usage:
```diff
@@ -4,4 +4,4 @@
 # Since the original script is system dependent and also requires /home/y location, this script is the dummy one
```
We should remove this script as well.

sschepens commented on conf/broker.conf:
```diff
@@ -212,10 +212,7 @@ loadBalancerReportUpdateThresholdPercentage=10
-
-# Frequency of sar report to collect
+# Frequency of report to collect
 loadBalancerHostUsageCheckIntervalMinutes=1
```
Does it really make sense to have intervals of < 1min? Would we be updating these reports to ZK everytime?

sschepens commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/BrokerHostUsage.java:
```diff
@@ -32,15 +41,26 @@
 public class BrokerHostUsage {
```
What interface do we plan on having? Just exposing `SystemResourceUsage getBrokerHostUsage()` would be enough?

merlimat replied:
Sounds good to me. Even some of the OS-indipendent methods (eg: used mem, processors count) could be put in a base abstract class.

sschepens commented:
@merlimat do these last changes seem ok? I've reused `loadManagerExecutor` to schedule usage calculations, and addressed most of the comments I believe.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/LinuxBrokerHostUsageImpl.java:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    private class CpuStat {
+        private Long totalTime;
```
We can use 2 `long` here

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java:
```diff
@@ -177,7 +177,7 @@ public SimpleLoadManagerImpl(PulsarService pulsar) {
         lastLoadReport = new LoadReport(pulsar.getWebServiceAddress(), pulsar.getWebServiceAddressTls(),
                 pulsar.getBrokerServiceUrl(), pulsar.getBrokerServiceUrlTls());
-        brokerHostUsage = new BrokerHostUsage(pulsar);
+        brokerHostUsage = new LinuxBrokerHostUsageImpl(pulsar);
```
This will generate errors when running on MacOS. We should have a basic factory that returns a generic host usage implementation when not running on Linux.

merlimat replied:
I mean, we don't need precise load information for MacOS, since it's not meant to be used for prod.. just not throwing exception or printing error logs, at least for now.

merlimat replied:
We could use the mbean based CPU usage for the "generic os"  implementation and keep using /proc/cpustat on Linux. 

For mac, we could omit he NIC traffic until we find a reasonable way to retrieve that.

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/LinuxBrokerHostUsageImpl.java:
```diff
@@ -0,0 +1,237 @@
+            return new CpuStat(total, total - idle);
+        } catch (IOException e) {
+            LOG.error("Failed to read CPU usage from /proc/stat", e);
+            return new CpuStat(0L, 0L);
```
what kind of usage % would we get when we fallback to `(0, 0)` ?
Would it look like 0% or 100% ?

merlimat replied:
I think we'd get "divide by 0 error" at line 88, since the CPU time diff would be 0

sschepens commented:
@merlimat pushed a generic implementation, could you review this again please?

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/GenericBrokerHostUsageImpl.java:
```diff
@@ -0,0 +1,88 @@
+        this.lastCollection = 0L;
+        this.totalCpuLimit = getTotalCpuLimit();
+        pulsar.getLoadManagerExecutor().scheduleAtFixedRate(this::checkCpuLoad, 0, CPU_CHECK_MILLIS, TimeUnit.MILLISECONDS);
+        pulsar.getLoadManagerExecutor().scheduleAtFixedRate(this::calculateBrokerHostUsage, 0, hostUsageCheckInterval, TimeUnit.SECONDS);
```
`hostUsageCheckInterval` is actually in minutes. Probably it'd be better to keep the time unit in the variable name to make it more explicit `hostUsageCheckIntervalMin`

sschepens replied:
sure

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/GenericBrokerHostUsageImpl.java:
```diff
@@ -0,0 +1,88 @@
+
+    private void checkCpuLoad() {
+        cpuUsageSum += systemBean.getSystemCpuLoad();
+        cpuUsageSum++;
```
`cpuUsageCount++`

merlimat commented on pulsar-broker/src/test/java/com/yahoo/pulsar/broker/loadbalance/SimpleLoadManagerImplTest.java:
```diff
@@ -441,8 +436,7 @@ public void testNamespaceBundleStats() {
     public void testBrokerHostUsage() {
-        when(pulsar1.getConfiguration().getLoadBalancerHostUsageScriptPath()).thenReturn("usageScript");
-        BrokerHostUsage brokerUsage = new BrokerHostUsage(pulsar1);
+        BrokerHostUsage brokerUsage = new LinuxBrokerHostUsageImpl(pulsar1);
```
This should either fallback to generic host usage impl or disable the test on non-linux.

sschepens commented:
@merlimat tests are green now, can you give this a last review?
